src/main/java/org/ebyhr/trino/storage/StorageMetadata.java

```diff
@@ -167,15 +154,69 @@ private ConnectorTableMetadata getStorageTableMetadata(ConnectorSession session,
             return null;
         }
 
-        return new ConnectorTableMetadata(tableName, table.getColumnsMetadata());
+        return new ConnectorTableMetadata(tableName.toSchemaTableName(), table.getColumnsMetadata());
+    }
+
+    /**
+     * Simplified variant of {@link SchemaTableName} that doesn't case-fold.
+     */
+    private static class SchemaTablePair
+    {
+        private String schemaName;
+        private String tableName;
+
+        public SchemaTablePair(String schemaName, String tableName)
+        {
+            this.schemaName = schemaName;
+            this.tableName = tableName;
+        }
+
+        public String getSchemaName()
+        {
+            return schemaName;
+        }
+
+        public String getTableName()
+        {
+            return tableName;
+        }
+
+        /**
+         * Convert to {@link SchemaTableName}
+         *
+         * NOTE: This will case-fold the schema- and table names, so they may no longer be possible to resolve afterwards.
+         */
+        public SchemaTableName toSchemaTableName()
+        {
+            return new SchemaTableName(schemaName, tableName);
+        }
     }
 
-    private List<SchemaTableName> listTables(ConnectorSession session, SchemaTablePrefix prefix)
+    private List<SchemaTablePair> listTables(ConnectorSession session, SchemaTablePrefix prefix)
     {
         if (prefix.getSchema().isPresent() && prefix.getTable().isPresent()) {
-            return List.of(new SchemaTableName(prefix.getSchema().get(), prefix.getTable().get()));
+            return List.of(new SchemaTablePair(prefix.getSchema().get(), prefix.getTable().get()));
+        }
+        return listTablesImpl(session, prefix.getSchema(), Function.identity());
```

Comment:
I think it would make sense to have `listTablesImpl` return a `Stream`, but we can't inline it since the Trino API expects it to return a `List<SchemaTableName>`.

Reply:
Unless you mean inline into `listTables(ConnectorSession, SchemaTablePrefix)`?

Reply:
Yes

Reply:
df2bdebd72878bed0e008025db591f304d3548f9